File: hik_driver/timer.py
from .utils import nanosec2date
import time

class Timer:
    Ksec2micro = 1000000

    def __init__(self, sync_point=10000000, freq=1):
        self.sync_point = sync_point
        self.freq = freq
        self.next_interval_us = None
        self.interval_us = int(Timer.Ksec2micro / freq)

    def get_current_time_us(self):
        # 获取当前时间的微秒时间戳
        current_time_us = int(time.time() * Timer.Ksec2micro)
        return current_time_us

    def sync_to_first_interval(self):
        current_time_us = self.get_current_time_us()
        sleep_time_us = self.sync_point - current_time_us % self.sync_point
        self.next_interval_us = current_time_us + sleep_time_us
        print(f"Wait until {nanosec2date(self.next_interval_us*1000)}")
        time.sleep(sleep_time_us / Timer.Ksec2micro - 0.021)

    def sync_to_next_interval(self):
        self.next_interval_us += self.interval_us
        sleep_t_us = self.next_interval_us - self.get_current_time_us()

        if sleep_t_us > 0:
            time.sleep(sleep_t_us / Timer.Ksec2micro - 0.000002)
        else:
            print("No sleep",flush=True)

File: src/timer.cpp
#include "timer.h"

Timer::Timer(uint64_t sync_point, double freq)
    : sync_point_(sync_point), freq_(freq)
{
    interval_us_ = static_cast<uint64_t>(Ksec2micro / freq_);
    next_interval_us_ = 0;
}

// 获取当前时间的微秒时间戳
uint64_t Timer::getCurrentTimeUs() const
{
    auto now = std::chrono::high_resolution_clock::now();
    auto now_us = std::chrono::duration_cast<std::chrono::microseconds>(
                      now.time_since_epoch())
                      .count();
    return now_us;
}

// 同步到第一个时间间隔
void Timer::syncToFirstInterval()
{
    uint64_t current_time_us = getCurrentTimeUs();
    uint64_t sleep_time_us = sync_point_ - (current_time_us % sync_point_);
    next_interval_us_ = current_time_us + sleep_time_us;

    std::cout << "Wait until " << nanosec2date(next_interval_us_ * 1000) << std::endl;

    // 睡眠
    std::this_thread::sleep_for(std::chrono::microseconds(sleep_time_us) -
                                std::chrono::microseconds(21000));
}

// 同步到下一个时间间隔
void Timer::syncToNextInterval()
{
    next_interval_us_ += interval_us_;
    int64_t sleep_t_us = next_interval_us_ - getCurrentTimeUs();

    if (sleep_t_us > 0)
    {
        std::this_thread::sleep_for(std::chrono::microseconds(sleep_t_us) -
                                    std::chrono::microseconds(2));
    }
    else
    {
        std::cout << "No sleep" << std::flush << std::endl;
    }
}